boxstream/unbox.go

```diff
@@ -14,62 +14,55 @@ import (
 // Unboxer decrypts everything that is read from it
 type Unboxer struct {
 	r      io.Reader
-	msg    []byte
+	buf    [MaxSegmentSize + secretbox.Overhead]byte
 	secret *[32]byte
 	nonce  *[24]byte
 }
 
-// readNextBox reads the next message from the underlying stream.
-func (u *Unboxer) readNextBox() error {
+// ReadMessage reads the next message from the underlying stream. If the next
+// message was a 'goodbye', it returns io.EOF.
+func (u *Unboxer) ReadMessage() ([]byte, error) {
```

Comment:
Right now `ReadMessage` always allocates a new `[]byte` to return, which is inefficient. One solution would be to return the `buf` slice directly, and mention in the docstring that its contents will change upon the next call to `ReadMessage`. Another common approach would be to use an "append-style" API, i.e. `AppendMessage(buf []byte) ([]byte, error)`, such that the caller manages the memory.